### PR TITLE
Checkstyle: enforce final keyword or encapsulation for class variables #319

### PR DESCRIPTION
**Pull request by eugenepeh**
Addition #319

### CODE REVIEW

**Comment by yamgent** [comment]
> On Hold
> have dependency on other PR regarding TestPerson rule violation

Ok, not sure how long the other PR will be on hold though (because I think it is waiting for another PR which is also on hold I think. :P). Feel free to resume this when the time is appropriate. :)

Changing to `s.OnHold`.

**Comment by eugenepeh** [comment]
Not sure if this is the best way to handle the public persons in TypicalPersons,
I personally feels that they should be changed to final if they were to remain public.

But there is also other workaround that can be used if this is not acceptable 
- Changing them to private with accessors 
- Suppress the error with filter
- Allow certain names by adding it to publicMemberPattern regex

**Comment by CanIHasReview** [comment]
# v1

@eugenepeh submitted v1 for review.

* [v1 1/1] [Checkstyle: add VisibilityModifier rule](https://github.com/se-edu/addressbook-level4/pull/364/commits/a5be2d2ed12ce2c12bfe40f6f1990eda830bd035)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6951a73c502185c54370c16fde85605743899d55...a5be2d2ed12ce2c12bfe40f6f1990eda830bd035))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/364/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 1]
```diff
@@ -62,4 +47,18 @@ public AddressBook getTypicalAddressBook() {
...
+    /**
+     * Internal person builder to handle IllegalValueException
+     */
+    private Person buildPerson(String name, String address, String email, String phone, String... tags) {
+        try {
+            return new PersonBuilder().withName(name).withAddress(address).withEmail(email).withPhone(phone)
+                    .withTags(tags).build();
+        } catch (IllegalValueException e) {
+            e.printStackTrace();
+            assert false : "not possible";
```
Comment (yamgent) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Actually, this is old style of doing assert. The new style is in this guideline: [Guidelines about Assertion](https://github.com/oss-generic/process/blob/master/codingStandards/AdditionalGuidelines-Java.md).

```
throw new AssertionError("This exception should not happen");
```

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 2]
```diff
@@ -62,4 +47,18 @@ public AddressBook getTypicalAddressBook() {
         loadAddressBookWithSampleData(ab);
         return ab;
     }
+
+    /**
+     * Internal person builder to handle IllegalValueException
+     */
+    private Person buildPerson(String name, String address, String email, String phone, String... tags) {
```
Comment (yamgent) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Was this method refactored out because of the change to `final`? If we change to throw `AssertionError` now, will this method still need to be refactored out?

Comment (eugenepeh) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Yes, this method is refactored out because of the change to final.
The main problem is due to the IllegalValueException, as Java does not let
the exception goes unhandled nor allows final declaration be inside a try catch,
Hence, I refactored it out to let the method handles the IllegalValueException.

**Comment by damithc** [comment]
@eugenepeh may be the PR title can be based on intent rather than mechanism.
e.g. `CheckStyle: enforce final keyword for ...`

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 3]
```diff
@@ -62,4 +47,18 @@ public AddressBook getTypicalAddressBook() {
...
+        try {
+            return new PersonBuilder().withName(name).withAddress(address).withEmail(email).withPhone(phone)
+                    .withTags(tags).build();
+        } catch (IllegalValueException e) {
+            e.printStackTrace();
```
Comment (damithc) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Instead of printing stack trace, you can use `AssertionError(String message, Throwable cause)` to create the exception object so that the stack trace is included inside the thrown exception object.

File: src/test/java/seedu/address/testutil/TypicalPersons.java (hunk 2 above)
Comment (eugenepeh) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: changed to throw AssertionError, looks good now and does not need to the new method anymore.
thanks @yamgent

**Comment by CanIHasReview** [comment]
# v2

@eugenepeh submitted v2 for review.

* [v2 1/1] [Checkstyle: add VisibilityModifier rule](https://github.com/se-edu/addressbook-level4/pull/364/commits/5eba573d670f9692ebd51e33ec647b3d7dbbcc00)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6951a73c502185c54370c16fde85605743899d55...5eba573d670f9692ebd51e33ec647b3d7dbbcc00)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/c87426b22f87558febf111a688fb10ccf92d9041~1...c87426b22f87558febf111a688fb10ccf92d9041))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/364/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 4]
```diff
@@ -38,8 +38,7 @@ public TypicalPersons() {
             ida = new PersonBuilder().withName("Ida Mueller").withPhone("8482131")
                     .withEmail("hans@example.com").withAddress("chicago ave").build();
         } catch (IllegalValueException e) {
-            e.printStackTrace();
-            assert false : "not possible";
+            throw new AssertionError("sample data cannot be invalid", e);
```
Comment (yamgent) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: `"sample...` -> `"Sample...`

Comment (eugenepeh) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Thanks @yamgent, I have copied this from other similar AssertionError. 
Should the others be rectified as well to make it standardize?

Comment (yamgent) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: The others can be rectified but should not be in this PR, instead a new `FirstTimer` issue should be created for it.

Comment (eugenepeh) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Thanks @yamgent, opened an issue #381 for it but I can't tag it. 😄

**Comment by CanIHasReview** [comment]
# v3

@eugenepeh submitted v3 for review.

* [v3 1/1] [Checkstyle: enforce final keyword or encapsulation for class variables](https://github.com/se-edu/addressbook-level4/pull/364/commits/a791b565d56c9fdaaf2431044517bd59f42afb4d)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/2c3390b1442d4fd7cc8484e339e9ffa3aed3b981...a791b565d56c9fdaaf2431044517bd59f42afb4d)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/d9497e075a29a06e5eda1d929a66107ee0657596~1...d9497e075a29a06e5eda1d929a66107ee0657596))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/364/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/guitests/AddressBookGuiTest.java [hunk 5]
```diff
@@ -53,7 +53,7 @@
     protected BrowserPanelHandle browserPanel;
     protected StatusBarFooterHandle statusBarFooter;
 
-    TestApp testApp;
+    private TestApp testApp;
```
Comment (damithc) [inline src/test/java/guitests/AddressBookGuiTest.java]: @eugenepeh is this a codacy bug?

Comment (eugenepeh) [inline src/test/java/guitests/AddressBookGuiTest.java]: @damithc this is due to the testApp only being used in the setup() method and not anywhere else

```
    @Before
    public void setup() throws Exception {
        FxToolkit.setupStage((stage) -> {
            mainGui = new MainGuiHandle(new GuiRobot(), stage);
            mainMenu = mainGui.getMainMenu();
            personListPanel = mainGui.getPersonListPanel();
            resultDisplay = mainGui.getResultDisplay();
            commandBox = mainGui.getCommandBox();
            browserPanel = mainGui.getBrowserPanel();
            statusBarFooter = mainGui.getStatusBarFooter();
            this.stage = stage;
        });
        EventsCenter.clearSubscribers();
        testApp = (TestApp) FxToolkit.setupApplication(() -> new TestApp(this::getInitialData, getDataFileLocation()));
        FxToolkit.showStage();
        while (!stage.isShowing());
        mainGui.focusOnMainApp();
    }
```

for the part where they didn't highlight the issue when TestApp is at package private level is maybe intended, but I will double check with codacy staff regarding this.

Comment (damithc) [inline src/test/java/guitests/AddressBookGuiTest.java]: @yamgent should we make `testApp` a local variable?

Comment (yamgent) [inline src/test/java/guitests/AddressBookGuiTest.java]: @damitch Sure. Since it is not used within the method as well, perhaps it doesn't even need to be a local variable? Just do `FxToolkit.setupApplication(() -> new TestApp(this::getInitialData, getDataFileLocation()));`

**Comment by CanIHasReview** [comment]
# v4

@eugenepeh submitted v4 for review.

* [v4 1/1] [Checkstyle: enforce final keyword or encapsulation for class variables](https://github.com/se-edu/addressbook-level4/pull/364/commits/679a5185831333355384373379bfbc29744caf6c)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/2c3390b1442d4fd7cc8484e339e9ffa3aed3b981...679a5185831333355384373379bfbc29744caf6c)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/05171637950a47d7e211142b05195fcad0293af5~1...05171637950a47d7e211142b05195fcad0293af5))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/364/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: config/checkstyle/checkstyle.xml [hunk 6]
```diff
@@ -237,6 +237,12 @@
 
     <module name="MissingSwitchDefault"/>
```
Comment (damithc) [inline config/checkstyle/checkstyle.xml]: The commit message body has no mention of `final` although the title mentions it.

**Comment by CanIHasReview** [comment]
# v5

@eugenepeh submitted v5 for review.

* [v5 1/1] [Checkstyle: enforce final keyword or encapsulation for class variables](https://github.com/se-edu/addressbook-level4/pull/364/commits/edc50c3986ee18a4bd24fbb488ff55567b71f150)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/2c3390b1442d4fd7cc8484e339e9ffa3aed3b981...edc50c3986ee18a4bd24fbb488ff55567b71f150)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/be2a8e2ec7dd802005142d6d1850607c3fdf938f~1...be2a8e2ec7dd802005142d6d1850607c3fdf938f))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/364/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 7]
```diff
@@ -38,8 +38,7 @@ public TypicalPersons() {
             ida = new PersonBuilder().withName("Ida Mueller").withPhone("8482131")
                     .withEmail("hans@example.com").withAddress("chicago ave").build();
         } catch (IllegalValueException e) {
-            e.printStackTrace();
-            assert false : "not possible";
+            throw new AssertionError("Sample data cannot be invalid", e);
```
Comment (damithc) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Remind me: why did we decide to include this change in this PR? Seems unrelated.

Comment (eugenepeh) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: This is due to Java does not let the exception goes unhandled nor allow final variable assignment to be inside a try catch. Hence upon changing the TestPersons to final, the IDE will prompt error such as `The blank final field ida may not have been initialized`

Originally I refactored a new method to handle the IllegalValueException. However, by changing to assertion to AssertionError exception, the new method becomes unnecessary since any error when assigning the value will cause the program to exits the method.

This is in a short discussion earlier [here](https://github.com/se-edu/addressbook-level4/pull/364#pullrequestreview-31072855)

Comment (damithc) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Thanks for the clarification @eugenepeh. I remember there was a reason, but couldn't remember it.
